x-pack/plugins/ml/public/file_datavisualizer/components/file_datavisualizer_view/file_datavisualizer_view.js

```diff
@@ -179,145 +193,94 @@ export class FileDataVisualizerView extends Component {
       fileTooLarge,
       fileCouldNotBeRead,
       serverErrorMessage,
+      mode,
     } = this.state;
 
     const fields = (results !== undefined && results.field_stats !== undefined) ? Object.keys(results.field_stats) : [];
 
     return (
       <React.Fragment>
-        <div style={{ textAlign: 'center' }} >
-          <EuiFilePicker
-            id="filePicker"
-            initialPromptText="Select or drag and drop a file"
-            onChange={files => this.onChange(files)}
-          />
-        </div>
-
-        <EuiSpacer size="l" />
-
-        {(loading) &&
-          <div style={{ textAlign: 'center' }} >
-            <EuiLoadingSpinner size="xl"/>
-          </div>
-        }
+        {(mode === MODE.READ) &&
+          <React.Fragment>
+            <div style={{ textAlign: 'center' }} >
+              <EuiFilePicker
+                id="filePicker"
+                initialPromptText="Select or drag and drop a file"
+                onChange={files => this.onChange(files)}
+              />
+            </div>
+
+            <EuiSpacer size="l" />
+
+            {(loading) &&
+              <div style={{ textAlign: 'center' }} >
+                <EuiLoadingSpinner size="xl"/>
+              </div>
+            }
 
-        {(fileTooLarge) &&
-          <FileTooLarge
-            fileSize={fileSize}
-            maxFileSize={MAX_BYTES}
-          />
-        }
+            {(fileTooLarge) &&
+              <FileTooLarge
+                fileSize={fileSize}
+                maxFileSize={MAX_BYTES}
+              />
+            }
 
-        {(fileCouldNotBeRead) &&
-          <FileCouldNotBeRead error={serverErrorMessage} />
-        }
+            {(fileCouldNotBeRead) &&
+              <FileCouldNotBeRead error={serverErrorMessage} />
+            }
 
-        {(loaded) &&
-          <ResultsView
-            results={results}
-            data={fileContents}
-            showEditFlyout={() => this.showEditFlyout()}
-          />
+            {(loaded) &&
+              <React.Fragment>
+                <ResultsView
+                  results={results}
+                  data={fileContents}
+                  showEditFlyout={() => this.showEditFlyout()}
+                />
+              </React.Fragment>
+            }
+            <EditFlyout
+              setShowFunction={this.setShowEditFlyoutFunction}
+              setOverrides={this.setOverrides}
+              originalSettings={this.originalSettings}
+              overrides={this.overrides}
+              fields={fields}
+            />
+
+            {(loaded /*&& results.format === 'delimited'*/) &&
+              <React.Fragment>
+                <EuiSpacer size="m" />
+                <EuiPanel>
+                  <EuiButton
+                    onClick={() => this.changeMode(MODE.IMPORT)}
```

Comment:
I think we should be careful about adding inline arrow functions in JSX - because the arrow function is recreated on each render it can potentially hurt performance and cause extra renders. A solution could be to create a wrapper function once on the class and then pass that to the component. Quick explanation here: https://medium.freecodecamp.org/why-arrow-functions-and-bind-in-reacts-render-are-problematic-f1c08b060e36

Reply:
hmm, this is an interesting point.
we do this a lot.
The impression i get from reading that article is it is bad to do when looping components.
This should be something we investigate further